algorithms/src/Kokkos_RadixSort.hpp

```diff
@@ -0,0 +1,251 @@
+        m_bits("radix_sort_bits", n) {}
+
+  // Generate and store the permutation induced by the keys, without
+  // modifying their initial order
+  template <class ExecutionSpace>
+  void create_indirection_vector(ExecutionSpace const& exec, View<T*> keys) {
+    auto key_functor = KeyFromView{keys};
+    const auto n     = keys.extent(0);
+
+    create_indirection_vector(exec, key_functor, n);
+  }
+
+  template <class ExecutionSpace, class KeyFunctor>
+  void create_indirection_vector(ExecutionSpace const& exec,
+                                 KeyFunctor key_functor, size_t n) {
+    RangePolicy<ExecutionSpace> policy(exec, 0, n);
+
+    // Initialize m_index_old, since it will be read from in the first
+    // iteration's call to step()
+    Kokkos::parallel_for(
+        policy, KOKKOS_LAMBDA(int i) { m_index_old(i) = i; });
+
+    for (int i = 0; i < num_bits; ++i) {
+      step<false>(policy, key_functor, i, m_index_new, m_index_old);
+      permute_by_scan<IndexType>(policy, {m_index_new, m_index_old});
+    }
+  }
+
+  // Directly re-arrange the entries of keys, optionally storing the permutation
+  template <bool store_permutation = false, class ExecutionSpace>
+  void sort(ExecutionSpace const& exec, View<T*> keys) {
+    // Almost identical to create_indirection_array, except actually permute the
+    // input
+    const auto n = keys.extent(0);
+    RangePolicy<ExecutionSpace> policy(exec, 0, n);
+
+    if constexpr (store_permutation) {
+      Kokkos::parallel_for(
+          policy, KOKKOS_LAMBDA(int i) { m_index_old(i) = i; });
+    }
+
+    for (int i = 0; i < num_bits; ++i) {
+      auto key_functor = KeyFromView{keys};
+
+      step<true>(policy, key_functor, i, m_index_new, m_index_old);
+      if constexpr (store_permutation) {
+        permute_by_scan<T, IndexType>(policy, {m_key_scratch, keys},
+                                      {m_index_new, m_index_old});
+      } else {
+        permute_by_scan<T>(policy, {m_key_scratch, keys});
+      }
+
+      // Number of bits is always even, and we know on odd numbered
+      // iterations we are reading from m_key_scratch and writing to keys
+      // So when this loop ends, keys will contain the results
+    }
+  }
+
+  // Directly re-arrange the entries of keys, optionally storing the permutation
+  template <bool store_permutation = false, class U, class ExecutionSpace>
+  void sortByKeys(ExecutionSpace const& exec, View<T*> keys, View<U*> values) {
+    // Almost identical to create_indirection_array, except actually permute the
+    // input
+    const auto n = keys.extent(0);
+    RangePolicy<ExecutionSpace> policy(exec, 0, n);
+
+    if constexpr (store_permutation) {
+      Kokkos::parallel_for(
+          policy, KOKKOS_LAMBDA(int i) { m_index_old(i) = i; });
+    }
+
+    auto values_scratch =
+        View<U*>(view_alloc(exec, "radix_sorter_values_scratch",
+                            Kokkos::WithoutInitializing),
+                 n);
```

Comment:
It would obviously be preferable to make this optionally persistent, so that it's not (de-)allocated repeatedly.